src/rgw/rgw_rados.cc

```diff
@@ -2291,38 +2293,70 @@ int RGWPutObjProcessor_Atomic::handle_data(bufferlist& bl, off_t ofs, MD5 *hash,
   *again = false;
 
   *phandle = NULL;
+
+  bufferlist in_bl;
+
+  // compression stuff
+  bool compression_enabled = store->ctx()->_conf->rgw_compression_type != "none";
+  if ((ofs > 0 && compressed) ||                                // if previous part was compressed
+      (ofs == 0 && compression_enabled)) {   // or it's the first part and flag is set
+    ldout(store->ctx(), 10) << "Compression for rgw is enabled, compress part" << dendl;
+    CompressorRef compressor = Compressor::create(store->ctx(), store->ctx()->_conf->rgw_compression_type);
+    if (!compressor.get()) {
+      // if compressor isn't available - just do not use it with log warning?
+      ldout(store->ctx(), 5) << "Cannot load compressor of type " << store->ctx()->_conf->rgw_compression_type 
+                       << "for rgw, check rgw_compression_type config option" << dendl;
+      compressed = false;
+      in_bl.claim(bl);
+    } else {
+      int cr = compressor->compress(bl, in_bl);
+      if (cr != 0) {
+        ldout(store->ctx(), 5) << "Compression failed with exit code " << cr << dendl;
+        compressed = false;
+        in_bl.claim(bl);
+      } else {
+        compressed = true;
```

Comment:
What's about handling the case when compression is not beneficial? Shouldn't we store uncompressed object?